diff --git a/drivers/char/tpm/tpm_ftpm_tee.c b/drivers/char/tpm/tpm_ftpm_tee.c
index 76adb108076c..301f19e945cc 100644
--- a/drivers/char/tpm/tpm_ftpm_tee.c
+++ b/drivers/char/tpm/tpm_ftpm_tee.c
@@ -15,10 +15,16 @@
 #include <linux/tee_drv.h>
 #include <linux/tpm.h>
 #include <linux/uuid.h>
+//#include <linux/atomic.h>
+#include <linux/delay.h>
 
 #include "tpm.h"
 #include "tpm_ftpm_tee.h"
 
+#include <linux/ima.h>
+#include <linux/rculist.h>
+#include "../../../security/integrity/ima/ima.h"
+
 /*
  * TA_FTPM_UUID: BC50D971-D4C9-42C4-82CB-343FB7F37896
  *
@@ -281,6 +287,54 @@ static int ftpm_tee_probe(struct device *dev)
 		goto out_chip;
 	}
 
+	atomic_inc(&ima_reinit); //increments the value of ima_reinit by 1
+	while(atomic_read(&_process_measurement) != 0){}; //busy-wait loop --> continuously checks the value of process_measurement until it becomes 0
+	ima_tpm_chip = chip;
+	ima_reinit_crypto();
+	ima_init_digests();
+
+	struct ima_queue_entry *qe;
+	struct ima_template_entry *entry;
+	u8 *digest;
+	struct tpm_digest *digests_arg;
+	int result;
+	u32 pcr = 10;  // Scegliere il PCR corretto
+	u16 alg_id;
+
+	list_for_each_entry(qe, &ima_measurements, later) {
+		entry = qe->entry;
+		for (int i = 0; i < NR_BANKS(ima_tpm_chip) + ima_extra_slots; i++) {
+			if (i == ima_sha1_idx)
+				continue;
+
+			if (i < NR_BANKS(ima_tpm_chip)) {
+				alg_id = ima_tpm_chip->allocated_banks[i].alg_id;
+				entry->digests[i].alg_id = alg_id;
+			}
+		}
+	}
+
+	list_for_each_entry(qe, &ima_measurements, later) {
+		entry = qe->entry;
+		digest = entry->digests[ima_hash_algo_idx].digest;
+		digests_arg = entry->digests;
+		result = tpm_pcr_extend(ima_tpm_chip, pcr, digests_arg);
+			if (result != 0)
+				pr_err("Error Communicating to TPM chip, result: %d\n", result);
+	}
+
+	atomic_set(&ima_reinit, 0); //sets the value of ima_reinit to 0
+	pr_info("tpm probe finished successfully\n");
+
+	if (chip->flags & TPM_CHIP_FLAG_TPM2)
+  		pr_info("optee-ftpm is a TPM 2.0\n");
+  
+	pr_info("optee-ftpm->nr_allocated_banks: %d\n", chip->nr_allocated_banks);
+
+	for (int i = 0; i < chip->nr_allocated_banks; i++) {
+		pr_info("optee-ftpm bank digest size: %d\n", chip->allocated_banks[i].digest_size);
+	}
+
 	return 0;
 
 out_chip:
diff --git a/security/integrity/ima/ima.h b/security/integrity/ima/ima.h
index 07a4586e129c..21bd4acf3ff4 100644
--- a/security/integrity/ima/ima.h
+++ b/security/integrity/ima/ima.h
@@ -21,9 +21,16 @@
 #include <linux/tpm.h>
 #include <linux/audit.h>
 #include <crypto/hash_info.h>
+#include <linux/atomic.h>
+#include <linux/wait.h> /* for wait_queue_head_t */
 
 #include "../integrity.h"
 
+extern atomic_t ima_reinit;
+extern atomic_t _process_measurement;
+extern struct tpm_digest *digests;
+extern struct ima_algo_desc *ima_algo_array;
+
 enum ima_show_type { IMA_SHOW_BINARY, IMA_SHOW_BINARY_NO_FIELD_LEN,
 		     IMA_SHOW_BINARY_OLD_STRING_FMT, IMA_SHOW_ASCII };
 enum tpm_pcrs { TPM_PCR0 = 0, TPM_PCR8 = 8, TPM_PCR10 = 10 };
@@ -50,10 +57,15 @@ extern int ima_policy_flag;
 extern atomic_t ima_setxattr_allowed_hash_algorithms;
 
 /* set during initialization */
-extern int ima_hash_algo __ro_after_init;
-extern int ima_sha1_idx __ro_after_init;
-extern int ima_hash_algo_idx __ro_after_init;
-extern int ima_extra_slots __ro_after_init;
+// extern int ima_hash_algo __ro_after_init;
+// extern int ima_sha1_idx __ro_after_init;
+// extern int ima_hash_algo_idx __ro_after_init;
+// extern int ima_extra_slots __ro_after_init;
+extern int ima_hash_algo;
+extern int ima_sha1_idx;
+extern int ima_hash_algo_idx;
+extern int ima_extra_slots;
+
 extern int ima_appraise;
 extern struct tpm_chip *ima_tpm_chip;
 extern const char boot_aggregate_name[];
@@ -109,6 +121,7 @@ struct ima_queue_entry {
 	struct ima_template_entry *entry;
 };
 extern struct list_head ima_measurements;	/* list of all measurements */
+extern struct list_head ima_new_measurements;	/* list of all measurements after ftpm init*/
 
 /* Some details preceding the binary serialized measurement list */
 struct ima_kexec_hdr {
@@ -149,6 +162,7 @@ void ima_add_violation(struct file *file, const unsigned char *filename,
 		       struct integrity_iint_cache *iint,
 		       const char *op, const char *cause);
 int ima_init_crypto(void);
+int ima_reinit_crypto(void);
 void ima_putc(struct seq_file *m, void *data, int datalen);
 void ima_print_digest(struct seq_file *m, u8 *digest, u32 size);
 int template_desc_init_fields(const char *template_fmt,
@@ -164,7 +178,7 @@ int ima_measurements_show(struct seq_file *m, void *v);
 unsigned long ima_get_binary_runtime_size(void);
 int ima_init_template(void);
 void ima_init_template_list(void);
-int __init ima_init_digests(void);
+int ima_init_digests(void);
 int ima_lsm_policy_change(struct notifier_block *nb, unsigned long event,
 			  void *lsm_data);
 
diff --git a/security/integrity/ima/ima_crypto.c b/security/integrity/ima/ima_crypto.c
index 51ad29940f05..a233d88f0672 100644
--- a/security/integrity/ima/ima_crypto.c
+++ b/security/integrity/ima/ima_crypto.c
@@ -62,15 +62,19 @@ struct ima_algo_desc {
 	enum hash_algo algo;
 };
 
-int ima_sha1_idx __ro_after_init;
-int ima_hash_algo_idx __ro_after_init;
+// int ima_sha1_idx __ro_after_init;
+// int ima_hash_algo_idx __ro_after_init;
+int ima_sha1_idx;
+int ima_hash_algo_idx;
 /*
  * Additional number of slots reserved, as needed, for SHA1
  * and IMA default algo.
  */
-int ima_extra_slots __ro_after_init;
+//int ima_extra_slots __ro_after_init;
+int ima_extra_slots;
 
-static struct ima_algo_desc *ima_algo_array;
+// static struct ima_algo_desc *ima_algo_array;
+struct ima_algo_desc *ima_algo_array;
 
 static int __init ima_init_ima_crypto(void)
 {
@@ -211,6 +215,113 @@ int __init ima_init_crypto(void)
 	return rc;
 }
 
+int ima_reinit_crypto(void)
+{
+    enum hash_algo algo;
+    long rc;
+    int i;
+
+	pr_err("ima reinit started successfully\n");
+    
+    /* Clear old ima_algo_array and associated tfm(s) */
+    for (i = 0; i < ima_extra_slots; i++) {
+        if (!ima_algo_array[i].tfm || ima_algo_array[i].tfm == ima_shash_tfm)
+            continue;
+
+        crypto_free_shash(ima_algo_array[i].tfm);
+    }
+    kfree(ima_algo_array);
+
+    ima_sha1_idx = -1;
+    ima_hash_algo_idx = -1;
+
+    for (i = 0; i < NR_BANKS(ima_tpm_chip); i++) {
+        algo = ima_tpm_chip->allocated_banks[i].crypto_id;
+        if (algo == HASH_ALGO_SHA1)
+            ima_sha1_idx = i;
+
+        if (algo == ima_hash_algo)
+            ima_hash_algo_idx = i;
+    }
+
+    if (ima_sha1_idx < 0) {
+        ima_sha1_idx = NR_BANKS(ima_tpm_chip) + ima_extra_slots++;
+        if (ima_hash_algo == HASH_ALGO_SHA1)
+            ima_hash_algo_idx = ima_sha1_idx;
+    }
+
+    if (ima_hash_algo_idx < 0)
+        ima_hash_algo_idx = NR_BANKS(ima_tpm_chip) + ima_extra_slots++;
+
+    ima_algo_array = kcalloc(NR_BANKS(ima_tpm_chip) + ima_extra_slots,
+                             sizeof(*ima_algo_array), GFP_KERNEL);
+    if (!ima_algo_array) {
+        rc = -ENOMEM;
+        goto out;
+    }
+
+    for (i = 0; i < NR_BANKS(ima_tpm_chip); i++) {
+        algo = ima_tpm_chip->allocated_banks[i].crypto_id;
+        ima_algo_array[i].algo = algo;
+
+        /* Unknown TPM algorithm */
+        if (algo == HASH_ALGO__LAST)
+            continue;
+
+        if (algo == ima_hash_algo) {
+            ima_algo_array[i].tfm = ima_shash_tfm;
+            continue;
+        }
+
+        ima_algo_array[i].tfm = ima_alloc_tfm(algo);
+        if (IS_ERR(ima_algo_array[i].tfm)) {
+            if (algo == HASH_ALGO_SHA1) {
+                rc = PTR_ERR(ima_algo_array[i].tfm);
+                ima_algo_array[i].tfm = NULL;
+                goto out_array;
+            }
+
+            ima_algo_array[i].tfm = NULL;
+        }
+    }
+
+    if (ima_sha1_idx >= NR_BANKS(ima_tpm_chip)) {
+        if (ima_hash_algo == HASH_ALGO_SHA1) {
+            ima_algo_array[ima_sha1_idx].tfm = ima_shash_tfm;
+        } else {
+            ima_algo_array[ima_sha1_idx].tfm = ima_alloc_tfm(HASH_ALGO_SHA1);
+            if (IS_ERR(ima_algo_array[ima_sha1_idx].tfm)) {
+                rc = PTR_ERR(ima_algo_array[ima_sha1_idx].tfm);
+                goto out_array;
+            }
+        }
+
+        ima_algo_array[ima_sha1_idx].algo = HASH_ALGO_SHA1;
+    }
+
+    if (ima_hash_algo_idx >= NR_BANKS(ima_tpm_chip) &&
+        ima_hash_algo_idx != ima_sha1_idx) {
+        ima_algo_array[ima_hash_algo_idx].tfm = ima_shash_tfm;
+        ima_algo_array[ima_hash_algo_idx].algo = ima_hash_algo;
+    }
+
+    return 0;
+
+out_array:
+    for (i = 0; i < NR_BANKS(ima_tpm_chip) + ima_extra_slots; i++) {
+        if (!ima_algo_array[i].tfm || ima_algo_array[i].tfm == ima_shash_tfm)
+            continue;
+
+        crypto_free_shash(ima_algo_array[i].tfm);
+    }
+    kfree(ima_algo_array);
+
+out:
+    crypto_free_shash(ima_shash_tfm);
+	
+    return rc;
+}
+
 static void ima_free_tfm(struct crypto_shash *tfm)
 {
 	int i;
diff --git a/security/integrity/ima/ima_init.c b/security/integrity/ima/ima_init.c
index 63979aefc95f..e096948ec912 100644
--- a/security/integrity/ima/ima_init.c
+++ b/security/integrity/ima/ima_init.c
@@ -20,6 +20,10 @@
 
 #include "ima.h"
 
+atomic_t ima_reinit=ATOMIC_INIT(0);
+atomic_t _process_measurement=ATOMIC_INIT(0);
+DECLARE_WAIT_QUEUE_HEAD(measurement_wq); // Declare a wait queue
+
 /* name for boot aggregate entry */
 const char boot_aggregate_name[] = "boot_aggregate";
 struct tpm_chip *ima_tpm_chip;
@@ -117,6 +121,9 @@ int __init ima_init(void)
 {
 	int rc;
 
+	atomic_set(&ima_reinit, 0);
+	atomic_set(&_process_measurement, 0);
+
 	ima_tpm_chip = tpm_default_chip();
 	if (!ima_tpm_chip)
 		pr_info("No TPM chip found, activating TPM-bypass!\n");
diff --git a/security/integrity/ima/ima_main.c b/security/integrity/ima/ima_main.c
index cc1217ac2c6f..9e9493a11855 100644
--- a/security/integrity/ima/ima_main.c
+++ b/security/integrity/ima/ima_main.c
@@ -35,7 +35,8 @@ int ima_appraise = IMA_APPRAISE_ENFORCE;
 int ima_appraise;
 #endif
 
-int __ro_after_init ima_hash_algo = HASH_ALGO_SHA1;
+// int __ro_after_init ima_hash_algo = HASH_ALGO_SHA1;
+int ima_hash_algo = HASH_ALGO_SHA1;
 static int hash_setup_done;
 
 static struct notifier_block ima_lsm_policy_notifier = {
@@ -223,8 +224,13 @@ static int process_measurement(struct file *file, const struct cred *cred,
 	enum hash_algo hash_algo;
 	unsigned int allowed_algos = 0;
 
-	if (!ima_policy_flag || !S_ISREG(inode->i_mode))
+	while(atomic_read(&ima_reinit) == 1){}; //busy-wait loop --> waits for ima_reinit to become 0 before proceeding
+	atomic_inc(&_process_measurement);
+
+	if (!ima_policy_flag || !S_ISREG(inode->i_mode)) {
+		atomic_dec(&_process_measurement);
 		return 0;
+	}
 
 	/* Return an IMA_MEASURE, IMA_APPRAISE, IMA_AUDIT action
 	 * bitmask based on the appraise/audit/measurement policy.
@@ -236,8 +242,10 @@ static int process_measurement(struct file *file, const struct cred *cred,
 	violation_check = ((func == FILE_CHECK || func == MMAP_CHECK ||
 			    func == MMAP_CHECK_REQPROT) &&
 			   (ima_policy_flag & IMA_MEASURE));
-	if (!action && !violation_check)
+	if (!action && !violation_check) {
+		atomic_dec(&_process_measurement);
 		return 0;
+	}
 
 	must_appraise = action & IMA_APPRAISE;
 
@@ -395,6 +403,7 @@ static int process_measurement(struct file *file, const struct cred *cred,
 				    pathname, "collect_data",
 				    "denied-hash-algorithm", rc, 0);
 	}
+	
 out_locked:
 	if ((mask & MAY_WRITE) && test_bit(IMA_DIGSIG, &iint->atomic_flags) &&
 	     !(iint->flags & IMA_NEW_FILE))
@@ -406,11 +415,14 @@ static int process_measurement(struct file *file, const struct cred *cred,
 	if (pathbuf)
 		__putname(pathbuf);
 	if (must_appraise) {
-		if (rc && (ima_appraise & IMA_APPRAISE_ENFORCE))
+		if (rc && (ima_appraise & IMA_APPRAISE_ENFORCE)) {
+			atomic_dec(&_process_measurement); //decrements process_measurement
 			return -EACCES;
+		}
 		if (file->f_mode & FMODE_WRITE)
 			set_bit(IMA_UPDATE_XATTR, &iint->atomic_flags);
 	}
+	atomic_dec(&_process_measurement); //decrements process_measurement
 	return 0;
 }
 
diff --git a/security/integrity/ima/ima_queue.c b/security/integrity/ima/ima_queue.c
index 532da87ce519..6f89edcd6e5b 100644
--- a/security/integrity/ima/ima_queue.c
+++ b/security/integrity/ima/ima_queue.c
@@ -22,9 +22,11 @@
 #define AUDIT_CAUSE_LEN_MAX 32
 
 /* pre-allocated array of tpm_digest structures to extend a PCR */
-static struct tpm_digest *digests;
+// static struct tpm_digest *digests;
+struct tpm_digest *digests;
 
 LIST_HEAD(ima_measurements);	/* list of all measurements */
+LIST_HEAD(ima_new_measurements);	/* list of all measurements after ftpm init*/
 #ifdef CONFIG_IMA_KEXEC
 static unsigned long binary_runtime_size;
 #else
@@ -211,7 +213,7 @@ int ima_restore_measurement_entry(struct ima_template_entry *entry)
 	return result;
 }
 
-int __init ima_init_digests(void)
+int ima_init_digests(void)
 {
 	u16 digest_size;
 	u16 crypto_id;
